crates/rattler_networking/src/oci_middleware.rs

```diff
@@ -0,0 +1,301 @@
+//! Middleware to handle `oci://` URLs to pull artifacts from an OCI registry
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use http::header::{ACCEPT, AUTHORIZATION};
+use reqwest::{Request, Response};
+use reqwest_middleware::{Middleware, Next, Result};
+use serde::Deserialize;
+use task_local_extensions::Extensions;
+use url::Url;
+
+use crate::mirror_middleware::create_404_response;
+
+/// Middleware to handle `oci://` URLs
+#[derive(Default, Debug, Clone)]
+pub struct OciMiddleware {
+    token_cache: Arc<Mutex<HashMap<Url, String>>>,
+}
+
+#[allow(dead_code)]
+enum OciAction {
+    Pull,
+    Push,
+    PushPull,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+struct OCIToken {
+    token: String,
+}
+
+impl ToString for OciAction {
+    fn to_string(&self) -> String {
+        match self {
+            OciAction::Pull => "pull".to_string(),
+            OciAction::Push => "push".to_string(),
+            OciAction::PushPull => "push,pull".to_string(),
+        }
+    }
+}
+// [oci://ghcr.io/channel-mirrors/conda-forge]/[osx-arm64/xtensor]
+async fn get_token(url: &Url, action: OciAction) -> Result<String> {
+    let token_url: String = format!(
+        "https://{}/token?scope=repository:{}:{}",
+        url.host_str().unwrap(),
+        &url.path()[1..],
+        action.to_string()
+    );
+
+    tracing::info!("Requesting token from {}", token_url);
+
+    let token = reqwest::get(token_url)
+        .await
+        .map_err(reqwest_middleware::Error::Reqwest)?
+        .json::<OCIToken>()
+        .await?
+        .token;
+
+    Ok(token)
+}
+
+fn oci_url_with_hash(url: &Url, hash: &str) -> Url {
+    format!(
+        "https://{}/v2{}/blobs/sha256:{}",
+        url.host_str().unwrap(),
+        url.path(),
+        hash
+    )
+    .parse()
+    .unwrap()
+}
+
+#[derive(Debug)]
+struct OciTagMediaType {
+    url: Url,
+    tag: String,
+    media_type: String,
+}
+
+#[allow(dead_code)]
```

Comment:
Is this function unused?

Reply:
Yeah it's currently not used. We would need it when we implement `push` to the OCI registry. We could also get rid of the code here for now.